ocaml/seahash/libseahash/seahash.c

```diff
@@ -0,0 +1,144 @@
+#include <stdint.h>
+#include <assert.h>
+#include <string.h>
+#include <stdio.h>
+
+/*
+	This code is a C implementation of the SeaHash algorithm (see: http://ticki.github.io/blog/seahash-explained/)
+	The author also provides a reference implemenation written in Rust (which follows the MIT license). 
+*/
+
+#define BUFF_SIZE (61440)
+
+uint64_t A, B, C, D;
+
+void init() {
+	// These default values come from the SeaHash specification
+	A = 0x16F11FE89B0D677C,
+	B = 0xB480A793D8E6C86C,
+	C = 0x6FE2E5AAF078EBC9,
+	D = 0x14F994A4C5259381;
+}
+
+uint64_t diffuse(uint64_t x) {
+	const uint64_t p = 0x6EED0E9DA4D94A4F;
+	x *= p;
+	x ^= (x >> 32) >> (x >> 60);
+	x *= p;
+	return x;
+}
+
+uint64_t finalise(const uint64_t size) {
+	A ^= B;
+	C ^= D;
+	A ^= C;
+	A ^= size;
+	return diffuse(A);
+}
+
+void rotate(uint64_t x) {
+	x = diffuse(A ^ x);
+	A = B; // Rotate the lanes
+	B = C;
+	C = D;
+	D = x;
+}
+
+void update(const void *data, const uint64_t size) {
+	const uint64_t freq = ((size >> 3) >> 2);
+	const uint64_t *head = (uint64_t *) data;
+
+	for(uint64_t i = 0; i < freq; i++) {
+		// NB: we assume little-endian, as we currently only support x86-64 systems
+		A ^= *head++;
+		B ^= *head++;
+		C ^= *head++;
+		D ^= *head++;
+
+		A = diffuse(A);
+		B = diffuse(B);
+		C = diffuse(C);
+		D = diffuse(D);
+	}
+
+	const uint8_t remain = size - (size & 0xFFFFFFFFFFFFFFE0); // How many bytes weren't hashed above (n < 32)
```

Comment:
Could this be expressed as `size & 0x1F` or similar? I'm not  comfortable with this subtraction.